### PR TITLE
Increased the test coverage of group_async_copy test.

### PR DESCRIPTION
**Pull request by JackAKirk**
vectors of size 2, 8 and 16 are now tested.
Updated deprecated calls to get_count() with size().

This PR is ready to be reviewed and considered for merging.

### CODE REVIEW

File: SYCL/Basic/group_async_copy.cpp [hunk 1]
```diff
@@ -153,6 +197,71 @@ int main() {
   for (int Stride = 1; Stride < WorkGroupSize; Stride++) {
     if (test<int>(Stride))
       return 1;
+    if (test<int2>(Stride))
+      return 1;
+    if (test<int4>(Stride))
+      return 1;
```
Comment (smaslov-intel) [inline SYCL/Basic/group_async_copy.cpp]: NIT: this is equivalent to the shorter
```suggestion
    if (test<int2>(Stride) ||
        test<int4>(Stride))
      return 1;
```

**Review by smaslov-intel — DISMISSED** [review]

File: SYCL/Basic/group_async_copy.cpp (hunk 1 above)
Comment (JackAKirk) [reply SYCL/Basic/group_async_copy.cpp]: Thanks, I've applied the suggestion generally.

**Review by smaslov-intel — APPROVED** [review]

**Review by smaslov-intel — DISMISSED** [review]

**Review by vladimirlaz — CHANGES_REQUESTED** [review]
Modified tets has failed on OpenCL Backed with the latest compiler:
```

[2022-02-28T10:13:27.157Z] ******************** TEST 'SYCL :: Basic/group_async_copy.cpp' FAILED ********************
[2022-02-28T10:13:27.157Z] Script:
[2022-02-28T10:13:27.157Z] --
[2022-02-28T10:13:27.157Z] : 'RUN: at line 1';    /runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/llvm.obj/bin/clang++      -fsycl -std=c++17 -fsycl-targets=spir64 /runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/llvm-test-suite/SYCL/Basic/group_async_copy.cpp -o /runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run
[2022-02-28T10:13:27.157Z] : 'RUN: at line 2';    env SYCL_DEVICE_FILTER=opencl:gpu,host  /runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run
[2022-02-28T10:13:27.157Z] : 'RUN: at line 3';   env SYCL_DEVICE_FILTER=opencl:cpu,host  /runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run
[2022-02-28T10:13:27.157Z] : 'RUN: at line 4';    env SYCL_DEVICE_FILTER=acc  /runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run
[2022-02-28T10:13:27.157Z] : 'RUN: at line 5';   env SYCL_DEVICE_FILTER=host /runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run
[2022-02-28T10:13:27.157Z] --
[2022-02-28T10:13:27.157Z] Exit Code: -6
[2022-02-28T10:13:27.157Z] 
[2022-02-28T10:13:27.157Z] Command Output (stdout):
[2022-02-28T10:13:27.157Z] --
[2022-02-28T10:13:27.157Z] $ ":" "RUN: at line 1"
[2022-02-28T10:13:27.157Z] note: command had no output on stdout or stderr
[2022-02-28T10:13:27.157Z] $ "/runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/llvm.obj/bin/clang++" "-fsycl" "-std=c++17" "-fsycl-targets=spir64" "/runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/llvm-test-suite/SYCL/Basic/group_async_copy.cpp" "-o" "/runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run"
[2022-02-28T10:13:27.157Z] note: command had no output on stdout or stderr
[2022-02-28T10:13:27.157Z] $ ":" "RUN: at line 2"
[2022-02-28T10:13:27.157Z] note: command had no output on stdout or stderr
[2022-02-28T10:13:27.157Z] $ "env" "SYCL_DEVICE_FILTER=opencl:gpu,host" "/runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run"
[2022-02-28T10:13:27.158Z] # command output:
[2022-02-28T10:13:27.158Z] Test passed.
[2022-02-28T10:13:27.158Z] 
[2022-02-28T10:13:27.158Z] $ ":" "RUN: at line 3"
[2022-02-28T10:13:27.158Z] note: command had no output on stdout or stderr
[2022-02-28T10:13:27.158Z] $ "env" "SYCL_DEVICE_FILTER=opencl:cpu,host" "/runDir/jenkins-dir/workspace/LLVM-Test-Suite-CI-TMP/LLVM-Test-Suite-CI-Linux/build/SYCL/Basic/Output/group_async_copy.cpp.tmp.run"
[2022-02-28T10:13:27.158Z] # command stderr:
[2022-02-28T10:13:27.158Z] terminate called after throwing an instance of 'cl::sycl::compile_program_error'
[2022-02-28T10:13:27.158Z]   what():  The program was built for 1 devices
[2022-02-28T10:13:27.158Z] Build program log for 'Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz':
[2022-02-28T10:13:27.158Z] Compilation started
[2022-02-28T10:13:27.158Z] Compilation done
[2022-02-28T10:13:27.158Z] Linking started
[2022-02-28T10:13:27.158Z] Linking done
[2022-02-28T10:13:27.158Z] Device build started
[2022-02-28T10:13:27.158Z] Options used by backend compiler: 
[2022-02-28T10:13:27.158Z] Failed to build device program
[2022-02-28T10:13:27.158Z] Error: unimplemented function(s) used:
[2022-02-28T10:13:27.158Z] _Z29async_work_group_strided_copyPU3AS3Dv2_DhPU3AS1KS_mm9ocl_event is undefined 
[2022-02-28T10:13:27.158Z] _Z29async_work_group_strided_copyPU3AS1Dv2_DhPU3AS3KS_mm9ocl_event is undefined 
[2022-02-28T10:13:27.158Z] _Z29async_work_group_strided_copyPU3AS3Dv4_DhPU3AS1KS_mm9ocl_event is undefined 
[2022-02-28T10:13:27.158Z] _Z29async_work_group_strided_copyPU3AS1Dv4_DhPU3AS3KS_mm9ocl_event is undefined 
[2022-02-28T10:13:27.158Z] _Z29async_work_group_strided_copyPU3AS3Dv8_DhPU3AS1KS_mm9ocl_event is undefined 
[2022-02-28T10:13:27.158Z] _Z29async_work_group_strided_copyPU3AS1Dv8_DhPU3AS3KS_mm9ocl_event is undefined 
[2022-02-28T10:13:27.158Z] CompilerException Failed to parse IR
[2022-02-28T10:13:27.158Z]  -11 (CL_BUILD_PROGRAM_FAILURE)
[2022-02-28T10:13:27.158Z] 
[2022-02-28T10:13:27.158Z] error: command failed with exit status: -6
[2022-02-28T10:13:27.158Z] 
[2022-02-28T10:13:27.158Z] --
```

**Comment by JackAKirk** [comment]
> Modified tets has failed on OpenCL Backed with the latest compiler:
> 

I think that this is because the test coverage has increased and apparently there is a problem with definitions/mangling of the double vector cases in the spir64 compilation that is using the generic implementation.  I looked into it and I don't see anything obviously wrong, the double vector cases have declarations and definitions.  Perhaps incorrect name mangling is being generated.

**Comment by JackAKirk** [comment]
> > Modified tets has failed on OpenCL Backed with the latest compiler:
> 
> I think that this is because the test coverage has increased and apparently there is a problem with definitions/mangling of the double vector cases in the spir64 compilation that is using the generic implementation. I looked into it and I don't see anything obviously wrong, the double vector cases have declarations and definitions. Perhaps incorrect name mangling is being generated.

@vladimirlaz 
I've investigated the non CUDA cases a bit more. I've noticed that the CI fails when I replace the deprecated `buffer.get_count()` with `buffer.size()` in addition to the above observation in my earlier comment.  I have "fixed" the failures by only applying the increased test coverage to the CUDA backend in a separate file. It seems that some of the increased coverage is not supported/needs to be fixed in other backends. This PR could now be merged to add the increased CUDA test coverage.  However some issues in other backends that this increased coverage has revealed could be investigated further.

Since I do not have rights to see the CI logs I tried to reproduce the problems on my local OpenCL CPU. Unfortunately I run into a different problem even for the (up to date) test that passes on the CI since it looks like I am not passing the test the correct parameters: I get:

```
InvalidModule: Invalid SPIR-V module: input SPIR-V module uses unknown extension 'SPV_KHR_expect_assume'

error: command failed with exit status: 9
```

**Comment by smaslov-intel** [comment]
@JackAKirk : what's the status of this? (looks stale)

**Comment by JackAKirk** [comment]
> @JackAKirk : what's the status of this? (looks stale)

Yes, it isn't really important. I'll close it.

**Pull request closed without merging**